volume region stats, csv reader, and cleanup / refactoring

petersteneteg opened the pull request:
Data Reader/Writer uses string_view
Exceptions has a new constructor for Eception(IVW_CONTEXT, <format_string>. args....)
MetaDataMap uses string_view
Added a Exporter base for processor that should export that to the regression, instead of just using CanvasProcessors
Added a LayerRamResize base, to handle layerRam resizing instead of abusing the data writer impl in Cimg
Filesystem more string_view uses
Raycasters: some documentation
Rewrite of the CSVReader, added CSVWriter
Added support for txt files, to the regression tests
Removed screenshots from the the regression since they consume loads of space
Made the DataFrameTable an Exporter to check CSV files in the regression
Added regression tests for the volume region stats

petersteneteg committed:
Volume: Added region stat processor

petersteneteg committed:
Volume: Made compiling the volumeregionstats 30x faster.

martinfalk commented on modules/volume/include/inviwo/volume/processors/volumeregionstatistics.h:
```diff
@@ -0,0 +1,70 @@
+
+/** \docpage{org.inviwo.VolumeRegionStatistics, Volume Region Statistics}
+ * ![](org.inviwo.VolumeRegionStatistics.png?classIdentifier=org.inviwo.VolumeRegionStatistics)
+ * Explanation of how to use the processor.
```
TODO

martinfalk commented on modules/volume/src/processors/volumeregionstatistics.cpp:
```diff
@@ -0,0 +1,385 @@
+    "Volume Region Statistics",           // Display name
+    "Undefined",                          // Category
+    CodeState::Experimental,              // Code state
+    Tags::None,                           // Tags
```
add category and tags

martinfalk commented on modules/volume/src/processors/volumeregionstatistics.cpp:
```diff
@@ -0,0 +1,385 @@
+
+        const auto posMin = dvec3{data2world * dvec4{0.0, 0.0, 0.0, 1.0}};
+        const auto posMax = dvec3{data2world * dvec4{1.0, 1.0, 1.0, 1.0}};
+        std::array<std::optional<dvec2>, 3> sizeRange = {{dvec2{posMin[0], posMax[0]},
```
unused?

petersteneteg replied:
no

martinfalk commented on modules/volume/src/processors/volumeregionstatistics.cpp:
```diff
@@ -0,0 +1,385 @@
+    }
+
+    size_t getRegion(size_t index) const {
+        return atlasRep->dispatch<size_t, dispatching::filter::UnsignedIntegerScalars>(
```
is that type guaranteed?

petersteneteg replied:
added a check

martinfalk commented on modules/volume/src/processors/volumeregionstatistics.cpp:
```diff
@@ -0,0 +1,385 @@
+            const auto p = indexMapper(pos);
+            for (size_t c = 0; c < channels; ++c) {
+                const auto region = regionMapper(getRegion(p) - 1, c);
+                const auto value = getValue(p, c);
```
Is it sensible to query the value n times instead of querying the entire vector once before the loop?

petersteneteg replied:
not sure if you mean the glm or the std vector... but In either case it seems it all gets inlined...

martinfalk replied:
Doesn't this call `getvalue()` in line 323, which uses a dispatch for each channel c. Maybe better to get the values for all channels and access them in the channels loop.

petersteneteg replied:
Had something like that at one point, but did not see a big difference, But have ot testet it for anything byt scalar volumes... if I do that I have to return a dvec4 from getValue even for scalar which might go the other way perf wise...

martinfalk reviewed (commented):
neat

petersteneteg committed:
Core/Tools/etc: Cleanup:

Data Reader/Writer uses string_view
Exceptions has a new constructor for Eception(IVW_CONTEXT, <format_string>. args....)
MetaDataMap uses string_view
Added a Exporter base for processor that should export that to the regression, instead of just using CanvasProcessors
Added a LayerRamResize base, to handle layerRam resizing instead of abusing the data writer impl in Cimg
Filesystem more string_view uses
Raycasters: some documentation
Rewrite of the CSVReader, added CSVWriter
Added support for txt files, to the regression tests
Removed screenshots from the the regression since they consume loads of space
Made the DataFrameTable an Exporter to check CSV files in the regression
Added regression tests for the volume region stats

petersteneteg committed:
Core: TmpFile bugfixes

petersteneteg committed:
Png: unittest bugfixes

petersteneteg committed:
Volume: regression removed unneeded test

Inviwo-Team committed:
Jenkins: Format fixes

petersteneteg committed:
Python: Fix, better error messages

petersteneteg committed:
Webbrowser: disable flaky tests

ResearchDaniel commented on include/inviwo/core/io/datareader.h:
```diff
@@ -81,7 +82,11 @@ class IVW_CORE_API DataReader {
     virtual std::any getOption([[maybe_unused]] std::string_view key) { return std::any{}; }
 
-private:
+protected:
```
This will make extensions_ open to inherited classes. Should add the private part below the functions

petersteneteg replied:
I don't see why that would be a problem

ResearchDaniel replied:
I don’t think it will be, more of a “keep as closed as possible” principle. It was like that before the change

petersteneteg replied:
Don't think “keep as closed as possible” makes any sense unless there is an invariant that you want to protect. And for the extensions_ I cant't think of any.

martinfalk commented on include/inviwo/core/processors/exporter.h:
```diff
@@ -0,0 +1,71 @@
+    /**
+     * Export some content to `path/name.ext` where ext is the first ext on candidateExtensions that
+     * is supported.
+     * @returns a string the the path to the exported file, of std::nullopt if no matching
```
-> @returns a string to the path of the exported file, or std::nullopt if no matching

ResearchDaniel replied:
returns nullopt if overwrite is set to No and a file exist?

ResearchDaniel replied:
Could also add info about if exceptions are thrown or captured by the function

martinfalk commented on include/inviwo/core/processors/exporter.h:
```diff
@@ -0,0 +1,71 @@
+
+namespace util {
+
+std::vector<std::string> exportAllFiles(ProcessorNetwork& network, std::string_view dir,
```
```suggestion
/**
 * Exports the data from all export processors in \p network into the directory \p dir using a \p nameTemplate 
 * and candidate extensions.
 * @return names of exported files
 */
std::vector<std::string> exportAllFiles(ProcessorNetwork& network, std::string_view dir,
```

martinfalk commented on modules/base/include/modules/base/io/datvolumesequencereader.h:
```diff
@@ -40,20 +40,34 @@ namespace inviwo {
+ *   - __SwizzleMask__  Data Channel swizzle mask @see SwizzleMask (Optional, defaults to rgba)
+ *   - __Interpolation__ Interpolation mode @see InterpolationType (Optional, defaults to Linear)
+ *   - __Wrapping__     Wrapping mode @see Wrapping3D (Optional defaults to clampAll
+ *   - __AxesNames__    Space separated axes names (Optional, defaults to "")
```
__AxisNames__ to be in line with __Axis1Name__?

martinfalk replied:
"Axis names" seems to be more common, see e.g. https://www.bigmountainanalytics.com/spotfire-axis-names-bar-charts/

martinfalk commented on modules/base/include/modules/base/io/datvolumesequencereader.h:
```diff
@@ -40,20 +40,34 @@ namespace inviwo {
+ *   - __Wrapping__     Wrapping mode @see Wrapping3D (Optional defaults to clampAll
+ *   - __AxesNames__    Space separated axes names (Optional, defaults to "")
+ *   - __Axis(1|2|3)Name__ Axis name (Optional, defaults to "")
+ *   - __AxesUnits__    Space separated axes units (Optional, defaults to Unit{})
```
be consistent with Axis/Axes Names

martinfalk commented on modules/base/include/modules/base/io/ivfsequencevolumewriter.h:
```diff
@@ -117,11 +118,10 @@ namespace util {
-                                                       std::string reltivePathToElements = "",
-                                                       bool overwrite = true);
+                                                       std::string_view name, std::string_view path,
+                                                       std::string_view reltivePathToElements = "",
```
```suggestion
                                                       std::string_view relativePathToElements = "",
```

martinfalk commented on modules/base/include/modules/base/io/datvolumesequencereader.h:
```diff
@@ -40,20 +40,34 @@ namespace inviwo {
+ *   - __ByteOffset__   Offset in to the raw file (Optional, default: 0)
+ *   - __DataOffset__   Offset in byte to where the data starts
+ *   - __Resolution | Dimension__ the size of the data grid: nx,ny,nz (Mandatory).
+ *   - __Format__       The type of values in the raw file. (Mandatory)
```
mention the valid formats as mentioned in the exception in datvolumesequencereader.cpp

martinfalk commented on modules/basegl/glsl/raycasting/raycaster-template.frag:
```diff
@@ -60,19 +59,27 @@ uniform float samplingRate = 2.0;
+    vec4 picking = vec4(0.0);  // The picking color of the ray
+    float rayDepth = -1.0;     // The ray depth value (0 to ray length, -1 mean "no" depth.
+                               // The same space as the rayPosition. Usually used to track
+                               // the depth of the "first" hit in DVR
```
```suggestion
    float rayDepth = -1.0;     // The ray depth value [0, ray length], -1 means "no" depth.
                               // Uses the same space as rayPosition. Usually used to track
                               // the depth of the "first" hit in DVR.
```

martinfalk commented on modules/basegl/glsl/raycasting/raycaster-template.frag:
```diff
@@ -60,19 +59,27 @@ uniform float samplingRate = 2.0;
+                               // the depth of the "first" hit in DVR
+    float depth = 1.0;         // The image depth, form far (0.0) to near (1.0).
+                               // Will be overridden by rayDepth if != -1 and
+                               // written to gl_FragDepth
```
```suggestion
    float depth = 1.0;         // The image depth, from far to near [0, 1].
                               // Will be overridden by rayDepth if != -1 and then
                               // written to gl_FragDepth
```

martinfalk commented on modules/basegl/glsl/raycasting/raycaster-template.frag:
```diff
@@ -60,19 +59,27 @@ uniform float samplingRate = 2.0;
+    //  * entryPointDepth the image depth of the entry point
+    //  * exitPointDepth the image depth of the exit point
+    //  * rayLength the distance from the start to the exit point in texture space
+    //  * rayDirection the direction of the ray in texture space, normalized.
```
```suggestion
    // The setup placeholder is expected to define:
    //  * entryPoint       the ray start point in the volume in texture coordinates
    //  * exitPoint        the ray exit point in the volume in texture coordinates
    //  * entryPointDepth  the image depth of the entry point
    //  * exitPointDepth   the image depth of the exit point
    //  * rayLength        the distance from the start to the exit point in texture space
    //  * rayDirection     the direction of the ray in texture space, normalized.
```

martinfalk commented on modules/cimg/src/cimgutils.cpp:
```diff
@@ -241,14 +244,14 @@ struct CImgNormalizedLayerDispatcher {
 
         try {
-            cimg_library::CImg<P> img(filePath.c_str());
+            cimg_library::CImg<P> img(SafeCStr{filePath}.c_str());
```
```suggestion
            cimg_library::CImg<P> img(SafeCStr<255>{filePath}.c_str());
```
file paths can be longer than the default 120 chars.

petersteneteg replied:
That is fine it will go to the heap then...

martinfalk commented on modules/dataframe/src/datastructures/column.cpp:
```diff
@@ -143,22 +126,23 @@ void CategoricalColumn::append(const Column& col) {
+        return it->second;
+    } else {
+        lookUpTable_.emplace_back(str);
+        auto ind = static_cast<glm::uint32_t>(lookUpTable_.size() - 1);
```
```suggestion
        auto ind = static_cast<std::uint32_t>(lookUpTable_.size() - 1);
```

martinfalk commented on modules/dataframe/src/io/csvreader.cpp:
```diff
@@ -33,262 +33,519 @@
 }
+CSVReader::EmptyField CSVReader::getHandleEmptyFields() const { return emptyField_; }
 
 bool CSVReader::setOption(std::string_view key, std::any value) {
```
this is a neat use case for `std::any`

martinfalk commented on modules/dataframe/src/io/csvreader.cpp:
```diff
@@ -33,262 +33,519 @@
+
+template <typename T>
+std::optional<T> toNumberLocale(std::string_view str) {
+    SafeCStr cstr{str};
```
are the default 120 chars enough? Probably better to put a size here.

martinfalk commented on modules/dataframe/src/io/csvreader.cpp:
```diff
@@ -33,262 +33,519 @@
+        char* end = nullptr;
+        typename std::remove_reference<decltype(errno)>::type errno_save = errno;
+        errno = 0;
+        int val = strtol(cstr.c_str(), &end, 10);
```
strtol casts to int without a warning?

petersteneteg replied:
at least on our systems long == int...

martinfalk commented on modules/dataframe/src/io/csvreader.cpp:
```diff
@@ -33,262 +33,519 @@
+    size_t sampledRows = sampleRows;
+    std::vector<CellType> types(nCol);
+
+    bool cLocale = locale_ == "C";
```
```suggestion
    const bool cLocale = locale_ == "C";
```

martinfalk commented on modules/dataframe/src/io/csvreader.cpp:
```diff
@@ -33,262 +33,519 @@
+    DataFrame& df, const std::vector<CellType>& types,
+    const std::vector<std::string>& headers) const {
+
+    bool cLocale = locale_ == "C";
```
```suggestion
    const bool cLocale = locale_ == "C";
```

martinfalk commented on modules/dataframe/src/io/csvwriter.cpp:
```diff
@@ -0,0 +1,138 @@
+    const std::array<char, 4> componentNames = {'X', 'Y', 'Z', 'W'};
+
+    // headers
+    auto oj = util::make_ostream_joiner(file, delimiter);
```
won't this always add a separator at the back?

petersteneteg replied:
no

martinfalk commented on modules/dataframe/src/io/csvwriter.cpp:
```diff
@@ -0,0 +1,138 @@
+
+    std::vector<std::function<void(std::ostream&, size_t)>> printers;
+    for (const auto& col : *dataFrame) {
+        if ((col == dataFrame->getIndexColumn()) && !exportIndexCol) {
```
```suggestion
        if ((col->getColumnType() == ColumnType::Index) && !exportIndexCol) {
```

martinfalk commented on modules/dataframe/src/io/csvwriter.cpp:
```diff
@@ -0,0 +1,138 @@
+            continue;
+        }
+        auto df = col->getBuffer()->getDataFormat();
+        if (auto cc = dynamic_cast<const CategoricalColumn*>(col.get())) {
```
```suggestion
        if (col->getColumnType() == ColumnType::Categorical) {
           auto cc = static_cast<const CategoricalColumn*>(col.get())
```

petersteneteg replied:
Think I will stick with the dynamic cast, there is no real perf aspect here

martinfalk commented on modules/dataframe/src/processors/csvsource.cpp:
```diff
@@ -52,23 +50,40 @@ CSVSource::CSVSource(const std::string& file)
+    , delimiters_("delimiters", "Delimiters", CSVReader::defaultDelimiters)
+    , stripQuotes_("stripQuotes", "Strip surrounding quotes", CSVReader::defaultStripQuotes)
+    , doublePrecision_("doublePrecision", "Double Precision", CSVReader::defaultDoublePrecision)
+    , exampleRows_("exampleRows", "Example Rows", CSVReader::defaultNumberOfExampleRows, 0, 10000)
```
what's the initial value? 100?

petersteneteg replied:
CSVReader::defaultNumberOfExampleRows think it might still be 50, but it will read more automatically

martinfalk commented on modules/dataframe/tests/unittests/csvreader-test.cpp:
```diff
@@ -347,11 +348,29 @@ TEST(CSVquotes, singlequotes) {
 }
 
-TEST(CVSquotes, headerquotes) {
+TEST(CSVquotes, StripSingleQuotes) {
```
```suggestion
TEST(CSVquotes, stripSingleQuotes) {
```

martinfalk commented on modules/dataframeqt/src/processors/dataframetable.cpp:
```diff
@@ -154,4 +155,15 @@ void DataFrameTable::setWidgetSize(size2_t dim) {
+        return util::saveData(*data, path, name, candidateExtensions, overwrite);
+    }
+
+    throw Exception("Could not find data frame", IVW_CONTEXT);
```
```suggestion
    throw Exception("Could not find DataFrame", IVW_CONTEXT);
```
alternative:
```suggestion
    throw Exception("Inport has no data", IVW_CONTEXT);
```

martinfalk commented on modules/nifti/src/niftireader.cpp:
```diff
@@ -162,14 +163,15 @@ const DataFormatBase* niftiDataTypeToInviwoDataFormat(const nifti_image* niftiIm
 
     /* read input dataset, but not data */
-    std::shared_ptr<nifti_image> niftiImage(nifti_image_read(filePath.c_str(), 0),
+    std::shared_ptr<nifti_image> niftiImage(nifti_image_read(SafeCStr{filePath}.c_str(), 0),
```
file paths can be longer than 120
```suggestion
    std::shared_ptr<nifti_image> niftiImage(nifti_image_read(SafeCStr<255>{filePath}.c_str(), 0),
```

petersteneteg replied:
thats fine, 120 is only for the stack

martinfalk commented on modules/plotting/src/utils/axisutils.cpp:
```diff
@@ -464,6 +464,10 @@ std::vector<std::pair<double, vec2>> getLabelPositions(const AxisSettings& setti
     std::vector<std::pair<double, vec2>> labelPositions(tickmarks.size());
 
+    if (axisLength > 1.0E10) {
+        LogInfoCustom("labelpos", "WTF" << axisLength << " " << settings.getCaption());
```
:-D
```suggestion
        LogInfoCustom("labelpos", "WTF " << axisLength << " " << settings.getCaption());
```

petersteneteg replied:
Ahh, did some very annoying Nan debugging there...

martinfalk commented on modules/png/src/pngreader.cpp:
```diff
@@ -42,46 +42,50 @@
     auto* fp = filesystem::fopen(filePath, "rb");
-    if (!fp) throw PNGLayerReaderException("Failed to open file for reading, " + filePath);
+    if (!fp)
+        throw DataReaderException(IVW_CONTEXT, "Failed to open file for reading, {}", filePath);
```
```suggestion
    if (!fp) {
        throw DataReaderException(IVW_CONTEXT, "Failed to open file for reading, {}", filePath);
    }
```

martinfalk commented on modules/png/src/pngreader.cpp:
```diff
@@ -42,46 +42,50 @@
+                                  "Internal PNG Error: Failed to create read struct: {}", name);
     }
-    util::OnScopeExit cleanup([&]() { png_destroy_read_struct(&png_ptr, NULL, NULL); });
+    util::OnScopeExit cleanup1([&]() { png_destroy_read_struct(&png_ptr, NULL, NULL); });
```
why `cleanup1` ?

petersteneteg replied:
since there is a cleanup2 below. you can not have them use the same OnScopeExit, that will delete them in thw wrong order. We have UB here before.

martinfalk commented on modules/pvm/src/pvmvolumereader.cpp:
```diff
@@ -77,19 +76,19 @@ std::shared_ptr<Volume> PVMVolumeReader::readPVMData(std::string filePath) {
-    unsigned char* pvmdata =
-        readPVMvolume(filePath.c_str(), &udim.x, &udim.y, &udim.z, &bytesPerVoxel, &spacing.x,
-                      &spacing.y, &spacing.z, &description, &courtesy, &parameter, &comment);
+    unsigned char* pvmdata = readPVMvolume(SafeCStr{filePath}.c_str(), &udim.x, &udim.y, &udim.z,
```
```suggestion
    unsigned char* pvmdata = readPVMvolume(SafeCStr<255>{filePath}.c_str(), &udim.x, &udim.y, &udim.z,
```

petersteneteg replied:
no need

martinfalk commented on modules/pvm/src/pvmvolumewriter.cpp:
```diff
@@ -98,38 +97,34 @@ void PVMVolumeWriter::writeData(const Volume* data, const std::string filePath)
     }
 
-    writePVMvolume(filePath.c_str(), dataPtr, static_cast<unsigned int>(dim.x),
+    writePVMvolume(SafeCStr{filePath}.c_str(), dataPtr, static_cast<unsigned int>(dim.x),
```
```suggestion
    writePVMvolume(SafeCStr<255>{filePath}.c_str(), dataPtr, static_cast<unsigned int>(dim.x),
```

petersteneteg replied:
no need

martinfalk commented on modules/volume/include/inviwo/volume/processors/volumeregionstatistics.h:
```diff
@@ -0,0 +1,86 @@
+ *   * __volume__ Segmented input volume
+ *   * __atlas__  Index volume, of unsigned integer type, assigning a region index to each voxel.
+ *                Has to have the same dimensions as volume. The index range is assumed to
+ *                be [0, dataMap.dataRange.y] and without gaps.
```
including 0 or not?

petersteneteg replied:
[] means inclusive, will have to come back to this at some later point though.

martinfalk replied:
just wanted to make sure that is what you meant. Not that 0 is some reserved value.

martinfalk commented on modules/volume/include/inviwo/volume/processors/volumeregionstatistics.h:
```diff
@@ -0,0 +1,86 @@
+ *   * __statistics__ Data Frame with the statistics for each region.
+ *
+ * ### Properties
+ *   * __Result Space__ The spacial domain of the resulting statistics. Data, Model, World, or
```
```suggestion
 *   * __Result Space__ The spatial domain of the resulting statistics. Data, Model, World, or
```

martinfalk commented on modules/volume/src/processors/volumeregionstatistics.cpp:
```diff
@@ -0,0 +1,402 @@
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+#include <tcb/span.hpp>
+#include <fmt/ostream.h>
```
```suggestion
```

petersteneteg replied:
?

martinfalk commented on src/core/common/inviwoapplication.cpp:
```diff
@@ -373,6 +374,9 @@ void InviwoApplication::setApplicationUsageMode(UsageMode mode) {
 }
 
+LayerRamResizer* InviwoApplication::getLayerRamResizer() const { return layerRamResizer_; }
+void InviwoApplication::setLayerRamResizer(LayerRamResizer* obj) { layerRamResizer_ = obj; }
```
```suggestion

void InviwoApplication::setLayerRamResizer(LayerRamResizer* obj) { layerRamResizer_ = obj; }
```

martinfalk commented on src/core/io/datawriter.cpp:
```diff
@@ -50,7 +53,24 @@ DataWriter& DataWriter::operator=(const DataWriter& that) {
+    checkOverwrite(path);
+    auto f = filesystem::ofstream(path, mode);
+    if (!f) {
+        throw FileException(IVW_CONTEXT, "could not open file '{}'", path);
```
```suggestion
        throw FileException(IVW_CONTEXT, "Could not open file '{}'", path);
```

martinfalk commented on src/core/util/filesystem.cpp:
```diff
@@ -93,35 +93,35 @@ std::string_view removeTrailingSlash(std::string_view path) {
     return std::fstream(util::toWstring(filename), mode);
 #else
-    return std::fstream(filename, mode);
+    return std::fstream(SafeCStr{filename}.c_str(), mode);
```
```suggestion
    return std::fstream(SafeCStr<255>{filename}.c_str(), mode);
```

martinfalk commented on src/core/util/filesystem.cpp:
```diff
@@ -93,35 +93,35 @@ std::string_view removeTrailingSlash(std::string_view path) {
     return std::ifstream(util::toWstring(filename), mode);
 #else
-    return std::ifstream(filename, mode);
+    return std::ifstream(SafeCStr{filename}.c_str(), mode);
```
```suggestion
    return std::ifstream(SafeCStr<255>{filename}.c_str(), mode);
```

martinfalk commented on src/core/util/filesystem.cpp:
```diff
@@ -93,35 +93,35 @@ std::string_view removeTrailingSlash(std::string_view path) {
     return std::ofstream(util::toWstring(filename), mode);
 #else
-    return std::ofstream(filename, mode);
+    return std::ofstream(SafeCStr{filename}.c_str(), mode);
```
```suggestion
    return std::ofstream(SafeCStr<255>{filename}.c_str(), mode);
```

martinfalk commented on src/qt/editor/inviwomainwindow.cpp:
```diff
@@ -385,7 +386,10 @@ void InviwoMainWindow::saveCanvases(std::string path, std::string fileName) {
 
-    util::saveAllCanvases(app_->getProcessorNetwork(), path, fileName);
+    util::exportAllFiles(
+        *app_->getProcessorNetwork(), path, fileName,
```
Would it be reasonable to make this available via the file menu as well? Like "Export workspace" or similar.

petersteneteg replied:
you can create a regression tests... that will save them...

petersteneteg replied:
and the workspace file...

martinfalk commented on tools/ivwpy/regression/reporttest.py:
```diff
@@ -55,30 +53,42 @@ def __init__(self, key, differenceTolerance=0.0):
+        tests = report[self.key]["tests"]
+        for test in tests:
+            if test['diff'] > 0:
+                self.message[test['txt']] = f"{test['txt']} differs from the reference" \
```
why `txt` ? Were there not also csv files?

petersteneteg replied:
txt mean any text based file, xml, csv, txt, anythnig that you can diff with a textual diff more or less

martinfalk commented on tools/ivwpy/regression/reporttest.py:
```diff
@@ -98,28 +108,32 @@ def test(self, report):
+                       "Missing reference text file"),
+            ReportTest(['txts', 'missing_txts'], lambda x: len(x) == 0,
+                       "Missing test text file"),
+            ReportTextTest('txts'),
```
csv files counting as `txts` files?

martinfalk reviewed (changes requested):
So many changes.
Maybe worthwhile putting some of those into the changelog.

ResearchDaniel commented on include/inviwo/core/common/inviwoapplication.h:
```diff
@@ -490,6 +495,7 @@ class IVW_CORE_API InviwoApplication : public Singleton<InviwoApplication> {
     WorkspaceManager::SerializationHandle presetsSerializationHandle_;
     WorkspaceManager::DeserializationHandle presetsDeserializationHandle_;
     std::unique_ptr<TimerThread> timerThread_;
+    LayerRamResizer* layerRamResizer_;
```
We should describe the approach for setting the layerRamResizer. That one should check that the currently set resizer is the owned one and if so set it to null upon deletion.

petersteneteg replied:
the modules usually own everything the app/factories only hold references. So I do the same here. the module is responsible for unsetting the resize when it is removed. 
If a module is unloaded one can not use the resizer even if it was owned by the app.

ResearchDaniel replied:
I changed my comment as I came to the same conclusion. Not sure if your comment is for the old one or the one I updated

petersteneteg replied:
added some comment with the getLayerRamResizer and setLayerRamResizer functions

ResearchDaniel commented on include/inviwo/core/datastructures/image/layerram.h:
```diff
@@ -184,4 +184,10 @@ auto LayerRAM::dispatch(Callable&& callable, Args&&... args) const -> Result {
+class IVW_CORE_API LayerRamResizer {
+public:
+    virtual ~LayerRamResizer() = default;
+    virtual bool resize(const LayerRAM& src, LayerRAM& dst) const = 0;
```
Should state the state of dst after resizing. 
If I am not mistaken the idea is that the contents of src is interpolated to fit into dst.

ResearchDaniel replied:
I think that this is still quite unclear. 
The layers do not change size, so there is no resize going on.
Come to think of it I think that it might be resolved if we call it LayerRamResampler.

ResearchDaniel commented on include/inviwo/core/io/datareaderexception.h:
```diff
@@ -32,14 +32,14 @@
 #include <inviwo/core/util/exception.h>
 
-#include <string>
+#include <string_view>
```
These includes could be removed?

petersteneteg replied:
yes, different PR...

ResearchDaniel commented on include/inviwo/core/io/datawriter.h:
```diff
@@ -52,13 +55,17 @@ class IVW_CORE_API DataWriter {
+    Overwrite getOverwrite() const;
+    void setOverwrite(Overwrite val);
+
+    static void checkOverwrite(std::string_view path, Overwrite overwrite);
```
What does checkOverwrite do? 
This needs to be clarified. 
After going through the code I suggest either adding documentation (it reads reasonably well in the code), or changing the name to something in line with:
```
    static void throwErrorOnOverwriteFail(std::string_view path, Overwrite overwrite);
```

The same goes for checkExists, which is perhaps more obvious. Better to be excplicit about the error throwing in my opinion.

petersteneteg replied:
I added some doc string

ResearchDaniel commented on include/inviwo/core/io/datawriterexception.h:
```diff
@@ -37,9 +37,7 @@ namespace inviwo {
-    DataWriterException(const std::string& message = "",
-                        ExceptionContext context = ExceptionContext());
-    virtual ~DataWriterException() noexcept = default;
+    using Exception::Exception;
```
Can remove the string include above now

petersteneteg replied:
different PR

ResearchDaniel commented on include/inviwo/core/io/datawriterutil.h:
```diff
@@ -0,0 +1,79 @@
+}
+
+template <typename T>
+std::optional<std::string> saveData(const T& data, std::string_view path, std::string_view name,
```
path is not entirely true and inconsistent with function above
```suggestion
std::optional<std::string> saveData(const T& data, std::string_view dir, std::string_view name,
```

petersteneteg replied:
changed the other one to filepath

petersteneteg replied:
and added comments

ResearchDaniel commented on include/inviwo/core/processors/exporter.h:
```diff
@@ -0,0 +1,71 @@
+/**
+ * \brief A base class for a Processor that might export a file. For example a CanvasProcessor
+ */
+class IVW_CORE_API Exporter {
```
The name is a bit broad for its purpose. Could potentially clash with something else in the inviwo namespace.

Perhaps put it in the util namespace along with the function below?

petersteneteg replied:
well we control what is in the Inviwo namespace so I don't worry so much about that.

ResearchDaniel commented on include/inviwo/core/util/exception.h:
```diff
@@ -73,39 +80,32 @@ std::basic_ostream<Elem, Traits>& operator<<(std::basic_ostream<Elem, Traits>& s
 public:
-    RangeException(std::string_view message = "", ExceptionContext context = ExceptionContext());
-    virtual ~RangeException() noexcept = default;
+    using Exception::Exception;
```
That's neat!

petersteneteg committed:
Various: review fixes

petersteneteg committed:
Volume: review fix

Co-authored-by: Martin Falk <martin.falk@liu.se>

ResearchDaniel commented on modules/base/src/io/ivfvolumewriter.cpp:
```diff
@@ -46,21 +47,16 @@ IvfVolumeWriter& IvfVolumeWriter::operator=(const IvfVolumeWriter& that) = defau
-    if (filesystem::fileExists(rawPath) && !overwrite)
-        throw DataWriterException("Output file: " + rawPath + " already exists",
-                                  IVW_CONTEXT_CUSTOM("util::writeIvfVolume"));
+    DataWriter::checkOverwrite(filePath, overwrite);
```
Copy-paste error
```suggestion
    DataWriter::checkExists(filePath, overwrite);
```

petersteneteg replied:
should have checked the raw file

Inviwo-Team committed:
Jenkins: Format fixes

ResearchDaniel commented on modules/basegl/glsl/raycasting/raycaster-template.frag:
```diff
@@ -60,19 +59,27 @@ uniform float samplingRate = 2.0;
+    float rayDepth = -1.0;     // The ray depth value (0 to ray length, -1 mean "no" depth.
+                               // The same space as the rayPosition. Usually used to track
+                               // the depth of the "first" hit in DVR
+    float depth = 1.0;         // The image depth, form far (0.0) to near (1.0).
```
```suggestion
    float depth = 1.0;         // The image depth, form far (0.0) to near (1.0).
                               // gl_FragDepth is set to rayDepth/rayLength if rayDepth != 1, otherwise depth.
```

ResearchDaniel commented on modules/basegl/include/modules/basegl/shadercomponents/sampletransformcomponent.h:
```diff
@@ -52,6 +53,7 @@ class IVW_MODULE_BASEGL_API SampleTransformComponent : public ShaderComponent {
     virtual std::vector<Segment> getSegments() override;
 
 private:
+    CompositeProperty sampleTransform_;
```
it is not really the sample (value) that is transformed? It is the position?
Might want to consider changing this to rayPosTransform/RayPosTransformComponent

petersteneteg replied:
In this context I don't think of sample as equivalent with value, the value is the result of the sample, you take a sample to get a value, and here I transform how that sample is taken.

ResearchDaniel commented on modules/basegl/src/processors/raycasting/atlasvolumeraycaster.cpp:
```diff
@@ -36,11 +36,11 @@ namespace inviwo {
-    CodeState::Experimental,                    // Code state
-    Tag::GL | Tag{"Volume"} | Tag{"Raycaster"}  // Tags
+    "org.inviwo.AtlasVolumeRaycaster",                         // Class identifier
+    "Atlas Volume Raycaster",                                  // Display name
```
Seems like Atlas is used in some places and Region is used in others?
We should make this consistent

petersteneteg replied:
yes, maybe a different PR. Not sure what it the bast names yet... Atlas/Region/Segment/... all have plus and minuses.

ResearchDaniel commented on modules/dataframe/include/inviwo/dataframe/datastructures/column.h:
```diff
@@ -313,6 +313,8 @@ class IVW_MODULE_DATAFRAME_API CategoricalColumn : public TemplateColumn<std::ui
 
     virtual void add(std::string_view value) override;
 
+    std::function<void(std::string_view)> addMany();
```
This deserves an explanation.

petersteneteg replied:
mmm

ResearchDaniel commented on modules/dataframe/include/inviwo/dataframe/io/csvreader.h:
```diff
@@ -60,27 +47,48 @@ class IVW_MODULE_DATAFRAME_API CSVDataReaderException : public DataReaderExcepti
+
+    enum class EmptyField { Throw, DefaultConstruct, NanOrZero };
+
+    CSVReader& setHandleEmptyFields(EmptyField emptyField);
```
Would be good to mention what it defaults to since it cannot be specified in the constructor

petersteneteg replied:
mm

ResearchDaniel commented on modules/dataframe/include/inviwo/dataframe/io/csvreader.h:
```diff
@@ -60,27 +47,48 @@ class IVW_MODULE_DATAFRAME_API CSVDataReaderException : public DataReaderExcepti
+    CSVReader& setLocale(std::string_view loc);
+    const std::string& getLocale() const;
+
+    enum class EmptyField { Throw, DefaultConstruct, NanOrZero };
```
NanOrZero? I think that these options should be separated

petersteneteg replied:
They already are, DefaultConstruct will generally be 0 for all types, NaNOrZero, sets NaNs for floating points and 0 for others. (for categorical ones the empty sting is not handled as a special case but just a an other value)

ResearchDaniel replied:
Ah, I think that from a user perspective we could use Zero, or EmptyOrZero, instead of DefaultConstruct then. While DefaultConstruct might make sense from a pure programmin perspective I think the other alternatives are more informative.
```suggestion
    enum class EmptyField { Throw, 
                            EmptyOrZero, ///< "" for strings/categorical and 0 for numerical values
                            NanOrZero    ///< "" for strings/categorical, NaN for floating point, and 0 for integer values
                                            };
```

ResearchDaniel commented on modules/dataframe/include/inviwo/dataframe/io/csvreader.h:
```diff
@@ -111,10 +119,40 @@ class IVW_MODULE_DATAFRAME_API CSVReader : public DataReaderType<DataFrame> {
+
 private:
+    struct CellType {
+        size_t integer;
```
nInteger
nReal
nString

petersteneteg replied:
I renamed the struct instead

ResearchDaniel commented on modules/dataframe/include/inviwo/dataframe/io/csvreader.h:
```diff
@@ -111,10 +119,40 @@ class IVW_MODULE_DATAFRAME_API CSVReader : public DataReaderType<DataFrame> {
+    bool unitsInHeaders_;
+    std::string unitRegexp_;
     bool doublePrecision_;
+    size_t exampleRows_;
```
nExampleRows_

petersteneteg replied:
seems excessive

ResearchDaniel commented on modules/dataframe/src/io/csvreader.cpp:
```diff
@@ -33,262 +33,519 @@
+             std::optional<size_t> refLine, Func&& func) {
+
+    std::string lookFor{delimiters};
+    lookFor.push_back('"');
```
Suggest to include an `empty` varable and use it in the code below to make it more explicit
```suggestion
    const auto empty = '"';
    lookFor.push_back(empty); 
```

petersteneteg replied:
it not empty, its a quote, added a comment.

ResearchDaniel commented on modules/dataframe/src/io/csvreader.cpp:
```diff
@@ -33,262 +33,519 @@
+            if (std::any_of(types.begin(), types.end(), [](const CellType& type) {
+                    return type.integer == 0 && type.real == 0 && type.string == 0;
+                })) {
+                sampledRows *= 2;
```
TImes 2?
Should be
`sampledRows = i`?

petersteneteg replied:
No, I increase the number sampled rows until we find non-empty ones.

ResearchDaniel commented on modules/dataframe/src/properties/columnmetadataproperty.cpp:
```diff
@@ -55,23 +56,27 @@ ColumnMetaDataProperty::ColumnMetaDataProperty(std::string_view identifier,
         .setCurrentStateAsDefault();
 
-    addProperties(header_, range_, unit_, drop_);
+    type_.setReadOnly(true);
```
Future PR: I think that it would be valid to change the type sometimes. E.g., to make something that was interpreted as integers to categorical. At least if the plotting behaves different for categorical compared to numerical columns

ResearchDaniel commented on modules/dataframe/tests/unittests/csvreader-test.cpp:
```diff
@@ -190,16 +190,16 @@ TEST(CSVdata, delimiterBeforeLineBreakWithHeader) {
-TEST(CSVdata, ignoreEmptyLine) {
-    // test whether rows with only delimiters, i.e. ",,,,," are ignored
+TEST(CSVdata, allEmptyElementLine) {
+    // test whether rows with only delimiters, i.e. ",,,,," are keept
```
```suggestion
    // test whether rows with only delimiters, i.e. ",,,,," are kept
```

ResearchDaniel commented on modules/opengl/include/modules/opengl/openglexception.h:
```diff
@@ -36,15 +36,13 @@ namespace inviwo {
 
 class IVW_MODULE_OPENGL_API OpenGLInitException : public ModuleInitException {
 public:
     OpenGLInitException(const std::string& message = "",
```
Cannot use?
`using ModuleInitException;`

petersteneteg replied:
out of scope

ResearchDaniel commented on modules/volume/include/inviwo/volume/processors/volumeregionstatistics.h:
```diff
@@ -0,0 +1,86 @@
+ *                      Index, defaults to World.
+ *
+ */
+class IVW_MODULE_VOLUME_API VolumeRegionStatistics : public PoolProcessor {
```
We mix atlas and region... We should use one of them

petersteneteg replied:
maybe I will rename the atlas renderer at some point.

ResearchDaniel commented on modules/volume/src/processors/volumeregionstatistics.cpp:
```diff
@@ -0,0 +1,402 @@
+        });
+
+        for (auto&& [i, stat] : util::enumerate(stats)) {
+            const auto [region, c] = regionMapper(i);
```
I think that iterating over regions would be quite useful in other places. If we would have an Atlas/Region volume it would be nice to expose such functionality there. Such a class could also ensure that the indices are consecutive and provide mappings to the original indices if need be

petersteneteg replied:
Im sure that could be nice, but the regionMapper here is just a regular IndexMapper2D.

ResearchDaniel reviewed (changes requested):
Nice!
Please separate such contributions into multiple PRs in the future.

petersteneteg commented:
I wanted to make smaller PRs but don't know how to do that in a effect manner. it is very hard to break these kinds of refactorings apart into smaller pieces that will still compile without a ton of trail and error. And even if you do you end up with a list of several PRs that all depend on each other so they have to be merged order, which menas you have to merge the first, push /rebase the next one, wait for CI /reviews, and then merge, then push the next... which ends up being really slow. If you have any suggestions for how to do that more efficiently that would be appreciated

petersteneteg committed:
Core/Dataframe: review fixes

petersteneteg committed:
DataFrame: doc fix

martinfalk commented on modules/dataframe/include/inviwo/dataframe/datastructures/column.h:
````diff
@@ -313,6 +315,24 @@ class IVW_MODULE_DATAFRAME_API CategoricalColumn : public TemplateColumn<std::ui
+     *     adder(str);
+     * }
+     * ```
+     */
````
```suggestion
     * For CategoricalColumn you cannot do the trick of pulling out the data container
     * `auto& data = templateColumn->getEditableRAMRepresentation()->getDataContainer();`
     * like for regular TemplateColumns to efficiently add many elements.
     * This is because adding element to the CategoricalColumn first has to find the string in the
     * lookup table and then insert the corresponding index into the data container. And we can't
     * do that from the "outside". This function returns a callable object `AddMany` that enables
     * this optimization for CategoricalColumns.
     * For example:
     * ```{.cpp}
     * CategoricalColumn col{...};
     * auto adder = col.addMany();
     * for (const auto& str : some_strings) {
     *     adder(str);
     * }
     * ```
     */
```

Inviwo-Team committed:
Jenkins: Format fixes

ResearchDaniel commented on include/inviwo/core/datastructures/image/layerram.h:
```diff
@@ -184,4 +184,15 @@ auto LayerRAM::dispatch(Callable&& callable, Args&&... args) const -> Result {
+     * Resize and copy src to dst doing some kind of content scaling
+     * @see CIMGLayerRamResizer
+     */
+    virtual bool resize(const LayerRAM& src, LayerRAM& dst) const = 0;
```
```suggestion
     /**
     * Resample the contents of src to dst while taking their potentially different sizes into account.
     * @see CIMGLayerRamResizer
     */
    virtual bool resample(const LayerRAM& src, LayerRAM& dst) const = 0;
```

martinfalk replied:
// copy the data from \p src to \p dst. This might involve up or down scaling if the dimensions do not match.
// The dimensions of both \p src and \p dst will not change.

ResearchDaniel reviewed (dismissed):
I would change the LayerRamResizer to LayerRamResampler
That would make its intent much clearer

ResearchDaniel commented:
> I wanted to make smaller PRs but don't know how to do that in a effect manner. it is very hard to break these kinds of refactorings apart into smaller pieces that will still compile without a ton of trail and error...

I understand that this evolved over time. My browser even freezes when github parses this many files though. I think that most string_view changes could be separated into a different PR, which would probably reduce the number of files by 2.

martinfalk commented on modules/dataframe/src/io/csvreader.cpp:
```diff
@@ -33,262 +33,519 @@
+#include <regex>
 
-namespace inviwo {
+#include <inviwo/core/util/clock.h>
```
remove?

martinfalk reviewed (changes requested):
Some minor things

petersteneteg committed:
Core/DataFrame: review updates

ResearchDaniel reviewed (approved):
Would still rename the resizer to resampler :)
